File: backend/app/models/group.py
"""Group models for user clubs and communities."""
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Integer, Text, ForeignKey, Boolean, Enum as SQLEnum
from sqlalchemy.orm import Mapped, mapped_column
import enum

from app.database import Base


class GroupPrivacy(enum.Enum):
    PUBLIC = "public"
    PRIVATE = "private"


class GroupRole(enum.Enum):
    OWNER = "owner"
    ADMIN = "admin"
    MEMBER = "member"


class Group(Base):
    """User group/club."""
    __tablename__ = "groups"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    slug: Mapped[str] = mapped_column(String(100), nullable=False, unique=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    privacy: Mapped[GroupPrivacy] = mapped_column(SQLEnum(GroupPrivacy), default=GroupPrivacy.PUBLIC)
    cover_image: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    icon: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    
    created_by: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    member_count: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)


class GroupMember(Base):
    """Group membership."""
    __tablename__ = "group_members"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    group_id: Mapped[int] = mapped_column(Integer, ForeignKey("groups.id"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    role: Mapped[GroupRole] = mapped_column(SQLEnum(GroupRole), default=GroupRole.MEMBER)
    joined_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)


class GroupCollection(Base):
    """Group content collection."""
    __tablename__ = "group_collections"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    group_id: Mapped[int] = mapped_column(Integer, ForeignKey("groups.id"), nullable=False, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_by: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    
    item_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)


class GroupCollectionItem(Base):
    """Item in a group collection."""
    __tablename__ = "group_collection_items"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    collection_id: Mapped[int] = mapped_column(Integer, ForeignKey("group_collections.id"), nullable=False, index=True)
    movie_id: Mapped[int] = mapped_column(Integer, ForeignKey("movies.id"), nullable=False, index=True)
    added_by: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    added_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)


class GroupDiscussion(Base):
    """Group discussion thread."""
    __tablename__ = "group_discussions"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    group_id: Mapped[int] = mapped_column(Integer, ForeignKey("groups.id"), nullable=False, index=True)
    author_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False)
    is_locked: Mapped[bool] = mapped_column(Boolean, default=False)
    
    reply_count: Mapped[int] = mapped_column(Integer, default=0)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)


class GroupDiscussionReply(Base):
    """Reply to a group discussion."""
    __tablename__ = "group_discussion_replies"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    discussion_id: Mapped[int] = mapped_column(Integer, ForeignKey("group_discussions.id"), nullable=False, index=True)
    author_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    content: Mapped[str] = mapped_column(Text, nullable=False)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)


class GroupActivity(Base):
    """Group activity feed entry."""
    __tablename__ = "group_activities"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    group_id: Mapped[int] = mapped_column(Integer, ForeignKey("groups.id"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    tenant_id: Mapped[int] = mapped_column(Integer, ForeignKey("tenants.id"), nullable=False, index=True)
    
    activity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)

File: frontend/src/api/group.ts
import apiClient from "./client";

export interface Group {
  id: number;
  name: string;
  slug: string;
  description?: string;
  privacy: string;
  member_count: number;
}

export interface GroupMember {
  user_id: number;
  role: string;
  joined_at: string;
}

export interface GroupCollection {
  id: number;
  name: string;
  description?: string;
  item_count: number;
}

export interface GroupCollectionItem {
  id: number;
  collection_id: number;
  movie_id: number;
  added_by: number;
  added_at: string;
}

export interface GroupDiscussion {
  id: number;
  title: string;
  author_id: number;
  reply_count: number;
  is_pinned: boolean;
  created_at: string;
}

export interface GroupDiscussionReply {
  id: number;
  discussion_id: number;
  author_id: number;
  content: string;
  created_at: string;
}

export interface GroupActivity {
  id: number;
  user_id: number;
  activity_type: string;
  content: string;
  created_at: string;
}

export interface CreateGroupRequest {
  name: string;
  slug: string;
  description?: string;
  privacy?: string;
}

export interface UpdateGroupRequest {
  name?: string;
  description?: string;
  privacy?: string;
}

export interface CreateCollectionRequest {
  name: string;
  description?: string;
}

export interface CreateDiscussionRequest {
  title: string;
  content: string;
}


export const createGroup = async (data: CreateGroupRequest): Promise<Group> => {
  const response = await apiClient.post<Group>("/groups", data);
  return response.data;
};

export const getGroups = async (): Promise<Group[]> => {
  const response = await apiClient.get<Group[]>("/groups");
  return response.data;
};

export const getGroup = async (id: number): Promise<Group> => {
  const response = await apiClient.get<Group>(`/groups/${id}`);
  return response.data;
};

export const updateGroup = async (id: number, data: UpdateGroupRequest): Promise<Group> => {
  const response = await apiClient.put<Group>(`/groups/${id}`, data);
  return response.data;
};

export const deleteGroup = async (id: number): Promise<void> => {
  await apiClient.delete(`/groups/${id}`);
};

export const joinGroup = async (id: number): Promise<void> => {
  await apiClient.post(`/groups/${id}/join`);
};

export const leaveGroup = async (id: number): Promise<void> => {
  await apiClient.post(`/groups/${id}/leave`);
};

export const inviteMember = async (groupId: number, userId: number): Promise<void> => {
  await apiClient.post(`/groups/${groupId}/invite`, { user_id: userId });
};

export const getMembers = async (groupId: number): Promise<GroupMember[]> => {
  const response = await apiClient.get<GroupMember[]>(`/groups/${groupId}/members`);
  return response.data;
};

export const updateMemberRole = async (groupId: number, userId: number, role: string): Promise<void> => {
  await apiClient.put(`/groups/${groupId}/members/${userId}`, { role });
};

export const removeMember = async (groupId: number, userId: number): Promise<void> => {
  await apiClient.delete(`/groups/${groupId}/members/${userId}`);
};

export const getCollections = async (groupId: number): Promise<GroupCollection[]> => {
  const response = await apiClient.get<GroupCollection[]>(`/groups/${groupId}/collections`);
  return response.data;
};

export const createCollection = async (groupId: number, data: CreateCollectionRequest): Promise<GroupCollection> => {
  const response = await apiClient.post<GroupCollection>(`/groups/${groupId}/collections`, data);
  return response.data;
};

export const addCollectionItem = async (groupId: number, collectionId: number, movieId: number): Promise<void> => {
  await apiClient.post(`/groups/${groupId}/collections/${collectionId}/items`, { movie_id: movieId });
};

export const removeCollectionItem = async (groupId: number, collectionId: number, itemId: number): Promise<void> => {
  await apiClient.delete(`/groups/${groupId}/collections/${collectionId}/items/${itemId}`);
};

export const getDiscussions = async (groupId: number): Promise<GroupDiscussion[]> => {
  const response = await apiClient.get<GroupDiscussion[]>(`/groups/${groupId}/discussions`);
  return response.data;
};

export const createDiscussion = async (groupId: number, data: CreateDiscussionRequest): Promise<GroupDiscussion> => {
  const response = await apiClient.post<GroupDiscussion>(`/groups/${groupId}/discussions`, data);
  return response.data;
};

export const createDiscussionReply = async (groupId: number, discussionId: number, content: string): Promise<GroupDiscussionReply> => {
  const response = await apiClient.post<GroupDiscussionReply>(`/groups/${groupId}/discussions/${discussionId}/replies`, { content });
  return response.data;
};

export const getActivityFeed = async (groupId: number): Promise<GroupActivity[]> => {
  const response = await apiClient.get<GroupActivity[]>(`/groups/${groupId}/activity`);
  return response.data;
};
